### PR TITLE
[function] Add a built-in function

### PR DESCRIPTION
**Pull request by qinkunbao**
## Description

Add a built-in function. 

Fixes # (issue)

## Type of change (select or add applied and delete the others)

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] API change with a documentation update
- [ ] Additional test coverage
- [ ] Code cleanup or just sync with upstream third-party crates

## Checklist

- [x] Fork the repo and create your branch from `master`.
- [x] If you've added code that should be tested, add tests.
- [x] If you've changed APIs, update the documentation.
- [x] Ensure the tests pass (see CI results).
- [x] Make sure your code lints/format.

### CODE REVIEW

**Comment by mssun** [comment]
Thanks.

You can add this example in the test by adding it here:

https://github.com/apache/incubator-teaclave/blob/51e7ee830dea022b87a758284d3ffb74cd63dadc/cmake/scripts/test.sh#L160

Then run `make run-examples` to test all examples.

**Comment by qinkunbao** [comment]
OK, I added this example and all tests passed.

**Comment by mssun** [comment]
Please use `make format` to format all rust code.

File: function/src/online_decrypt.rs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+enum Error {
+    CryptoError,
+}
+
+impl From<ring::error::Unspecified> for Error {
+    fn from(_: ring::error::Unspecified) -> Self {
+        Error::CryptoError
+    }
+}
```
Comment (mssun) [inline function/src/online_decrypt.rs:46]: These are no longer used.

File: function/src/online_decrypt.rs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    key: &str,
+    nonce_str: &str,
+    encrypted: &str,
+    alg: &'static ring::aead::Algorithm,
+) -> anyhow::Result<String> {
+    let decoded_key = base64::decode(&key)?;
+    let nonce = base64::decode(&nonce_str)?;
+    let mut data_vec = base64::decode(&encrypted)?;
+
+    let _result = decrypt(&decoded_key, &nonce, &mut data_vec, &alg)
```
Comment (mssun) [inline function/src/online_decrypt.rs:83]: `     let _result = decrypt(&decoded_key, &nonce, &mut data_vec, &alg) ` => `decrypt(&decoded_key, &nonce, &mut data_vec, &alg) `

No need to get the `()` return.

File: executor/Cargo.toml [hunk 3]
```diff
@@ -30,13 +30,16 @@ full_builtin_function = [
...
+  "builtin_logistic_regression_train",
+  "builtin-online-decrypt",
 ]
+
 builtin_echo = []
 builtin_gbdt_predict = []
 builtin_gbdt_train = []
 builtin_logistic_regression_predict = []
 builtin_logistic_regression_train = []
+builtin-online-decrypt = []
```
Comment (mssun) [inline executor/Cargo.toml:42]: Use `_` instead of `-` for consistency.

File: executor/Cargo.toml [hunk 4]
```diff
@@ -30,13 +30,16 @@ full_builtin_function = [
   "builtin_gbdt_predict",
   "builtin_gbdt_train",
   "builtin_logistic_regression_predict",
-  "builtin_logistic_regression_train"
+  "builtin_logistic_regression_train",
+  "builtin-online-decrypt",
```
Comment (mssun) [inline executor/Cargo.toml:34]: Use `_` instead of `-` for consistency.

File: examples/python/builtin_online_decrypt.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    ]
+    nonce = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
+    encrypted_data = "CaZd8qSMMlBp8SjSXj2I4dQIuC9KkZ5DI/ATo1sWJw=="
+    algorithm = "aes256gcm" #Alogorithm can be "aes256gcm" or "aes128gcm"
+    rt = example.decrypt(key, nonce, encrypted_data, algorithm)
+    print("[+] function return: ", rt)
+
+
+if __name__ == '__main__':
+    main()
```
Comment (mssun) [inline examples/python/builtin_online_decrypt.py:84]: Missing newline in the end of the file.

File: examples/python/builtin_online_decrypt.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+def main():
+    example = BuiltinOnlineDecryptExample(USER_ID, USER_PASSWORD)
+    key = [
+        106, 165, 29, 129, 157, 37, 38, 123, 179, 247, 40, 143, 146, 128, 241,
+        51, 166, 92, 77, 197, 85, 165, 222, 10, 40, 186, 179, 108, 112, 252,
+        240, 184
+    ]
+    nonce = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
+    encrypted_data = "CaZd8qSMMlBp8SjSXj2I4dQIuC9KkZ5DI/ATo1sWJw=="
+    algorithm = "aes256gcm" #Alogorithm can be "aes256gcm" or "aes128gcm"
```
Comment (mssun) [inline examples/python/builtin_online_decrypt.py:78]: The comment sign `#` comes with a space like this:

`#Alogorithm can be "aes256gcm" or "aes128gcm"` => `# Alogorithm can be "aes256gcm" or "aes128gcm"`

File: examples/python/builtin_online_decrypt.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        client.invoke_task(task_id)
+
+        print("[+] getting result")
+        result = client.get_task_result(task_id)
+        print("[+] done")
+
+        return bytes(result)
+
+
+def DataToBase64(data):
```
Comment (mssun) [inline examples/python/builtin_online_decrypt.py:65]: Function in Python should be in camelCase.

`DataToBase64` => `dataToBase64`

File: executor/src/builtin.rs [hunk 8]
```diff
@@ -49,6 +49,8 @@ impl TeaclaveExecutor for BuiltinFunctionExecutor {
             LogisticRegressionPredict::NAME => {
                 LogisticRegressionPredict::new().run(arguments, runtime)
             }
+            #[cfg(feature = "builtin-online-decrypt")]
```
Comment (mssun) [inline executor/src/builtin.rs:52]: Use `_` instead of `-` for consistency.

File: function/src/online_decrypt.rs [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        let args = OnlineDecryptArguments::try_from(arguments)?;
+        let key = args.key;
+        let nonce = args.nonce;
+        let encrypted_data = args.encrypted_data;
+        let algorithm = &args.algorithm[..];
+
+        let alg = match algorithm {
+            "aes128gcm" => &AES_128_GCM,
+            "aes256gcm" => &AES_256_GCM,
+            _ => return Err(anyhow!("Invalid algorithm")),
```
Comment (mssun) [inline function/src/online_decrypt.rs:111]: `return Err(anyhow!("Invalid algorithm"))` == `bail!("Invalid algorithm")`

File: function/src/online_decrypt.rs (hunk 1 above)
Comment (mssun) [reply function/src/online_decrypt.rs:46]: Please delete them.

**Comment by qinkunbao** [comment]
Thank you for the feedback. All requested changes are addressed now.

**Comment by mssun** [comment]
Merged. Thanks.

**Pull request merged**